File: app/src/main/java/com/microlabs/trallet/adapter/ExpenseRVAdapter.java
package com.microlabs.trallet.adapter;

import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.microlabs.trallet.AddExpensesActivity;
import com.microlabs.trallet.ExpenseActivity;
import com.microlabs.trallet.R;
import com.microlabs.trallet.base.RealmHelper;
import com.microlabs.trallet.model.Expense;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by Nicholas on 3/13/2017.
 */

public class ExpenseRVAdapter extends RecyclerView.Adapter<ExpenseRVAdapter.ViewHolder> {

    Context context;
    List<Expense> expenseList = new ArrayList<>();

    public ExpenseRVAdapter(Context context) {
        this.context = context;
    }

    public void updateData(List<Expense> expenses) {
        expenseList.clear();
        expenseList.addAll(expenses);
        notifyDataSetChanged();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = View.inflate(context, R.layout.item_expense, null);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        Expense expense = expenseList.get(position);
        holder.updateItem(expense);
    }

    @Override
    public int getItemCount() {
        return expenseList.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.txtExTitle)
        TextView txtExTitle;
        @BindView(R.id.txtExPrice)
        TextView txtExPrice;
        Expense item;

        ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        void updateItem(Expense item) {
            this.item = item;
            txtExTitle.setText(item.getTitle());
            txtExPrice.setText(String.valueOf(item.getValue()));
        }

        @OnClick({R.id.btnEdit, R.id.btnDelete, R.id.cvExpense})
        public void onClick(View view) {
            switch (view.getId()) {
                case R.id.btnEdit:
                    Intent intent = new Intent(context, AddExpensesActivity.class);
                    intent.putExtra("bookId", item.getBookId());
                    intent.putExtra("fromAdapter", item.getId());
                    context.startActivity(intent);
                    break;
                case R.id.btnDelete:
                    ((ExpenseActivity) context).deleteExpense(item.getBookId(), item.getValue());
                    RealmHelper.deleteExpense(item.getId());
                    ((ExpenseActivity) context).setUpView();
                    break;
                case R.id.cvExpense:
                    break;
            }
        }
    }
}

File: app/src/main/java/com/microlabs/trallet/adapter/ExpenseRVAdapter.kt
package com.microlabs.trallet.adapter

import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.microlabs.trallet.ExpenseActivity
import com.microlabs.trallet.R
import com.microlabs.trallet.model.Expense
import kotlinx.android.synthetic.main.item_expense.view.*
import java.util.*

/**
 * Created by Nicholas on 3/13/2017.
 *
 *
 * Expense RecyclerView Adapter
 */

class ExpenseRVAdapter(internal var itemListener: ExpenseActivity.ExpenseItemListener) : RecyclerView.Adapter<ExpenseRVAdapter.ViewHolder>() {
    private val expenseList = ArrayList<Expense>()

    fun updateData(expenses: List<Expense>) {
        expenseList.clear()
        expenseList.addAll(expenses)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(View.inflate(parent.context, R.layout.item_expense, null))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.updateItem(expenseList[position])
    }

    override fun getItemCount(): Int {
        return expenseList.size
    }

    fun removeExpense(position: Int) {
        expenseList.removeAt(position)
        notifyItemRemoved(position)
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        lateinit var item: Expense
        val txtExTitle = itemView.txtExTitle
        val txtExPrice = itemView.txtExPrice
        val imgExpense = itemView.imgExpense
        val btnEdit = itemView.btnEdit
        val btnDelete = itemView.btnDelete
        val cvExpense = itemView.cvExpense

        internal fun updateItem(item: Expense) {
            this.item = item
            txtExTitle.text = item.title
            txtExPrice.text = item.value.toString()
            when (item.categoryId) {
                1 -> imgExpense.setImageResource(R.drawable.ic_local_dining_black_24dp)
                2 -> imgExpense.setImageResource(R.drawable.ic_local_taxi_black_24dp)
                3 -> imgExpense.setImageResource(R.drawable.ic_shopping_basket_black_24dp)
                else -> imgExpense.setImageResource(R.drawable.ic_local_atm_black_24dp)
            }

            btnEdit.setOnClickListener {
                itemListener.onEditClick(item.id)
            }
            btnDelete.setOnClickListener {
                itemListener.onDeleteClick(item)
            }
            cvExpense.setOnClickListener {
                itemListener.onEditClick(item.id)
            }
        }
    }
}
